File: javaCalc/src/main/java/calculator/operations/AbstractBasicOperation.java
package calculator.operations;

import java.util.Objects;

public abstract class AbstractBasicOperation implements Operation {
    public final static int PARAMETERS_COUNT = 2;
    private final String operator;

    AbstractBasicOperation(String operator) {
        this.operator = operator;
    }

    @Override
    public String getOperator() {
        return operator;
    }

    @Override
    public Integer apply(Integer... numbers) {
        checkParametersCount(numbers.length);
        return apply(numbers[0], numbers[1]);
    }

    protected abstract Integer apply(Integer a, Integer b);

    public void checkParametersCount(Integer parametersCount) {
        if (parametersCount != PARAMETERS_COUNT) {
            throw new IllegalArgumentException(String.format(
                    "неверное колличество параметров, операции '%s' требуется %d параметров",
                    getOperator(),
                    PARAMETERS_COUNT)
            );
        }
    }

    @Override
    public String toString() {
        return operator;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AbstractBasicOperation that = (AbstractBasicOperation) o;
        return Objects.equals(operator, that.operator);
    }

    @Override
    public int hashCode() {
        return Objects.hash(operator);
    }
}

File: kotlinCalc/src/main/kt/calculator/operations/AbstractBasicOperation.kt
package calculator.operations

import java.util.*

abstract class AbstractBasicOperation(override val operator: String) : Operation {
    override fun apply(numbers: List<Int>): Int {
        checkParametersCount(numbers.size)
        return apply(numbers[0], numbers[1])
    }

    protected open fun apply(a: Int, b: Int): Int {
        return a / b
    }

    override fun checkParametersCount(parametersCount: Int) {
        require(parametersCount == PARAMETERS_COUNT) {
            """неверное колличество параметров, операции '$operator' требуется $PARAMETERS_COUNT параметров"""
        }
    }

    override fun toString(): String {
        return operator
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as AbstractBasicOperation
        return operator == that.operator
    }

    override fun hashCode(): Int {
        return Objects.hash(operator)
    }

    companion object {
        var PARAMETERS_COUNT = 2
    }
}